sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CollationTypeCoercion.scala

```diff
@@ -461,6 +421,66 @@ object CollationTypeCoercion {
         else right
     }
   }
+
+  /**
+   * Throws an analysis exception if the new data type has indeterminate collation,
+   * and the expression is not allowed to have inputs with indeterminate collations.
+   */
+  private def checkIndeterminateCollation(expression: Expression, newDataType: DataType): Unit = {
+    if (shouldFailWithIndeterminateCollation(expression, newDataType)) {
+      expression.failAnalysis(
+        errorClass = "INDETERMINATE_COLLATION_IN_EXPRESSION",
+        messageParameters = Map("expr" -> toSQLExpr(expression)))
+    }
+  }
+
+  /**
+   * Returns whether the given expression which isn't allowed to have inputs with indeterminate
```

Comment:
Consider rewording the comment to:
Returns whether the given expression has indeterminate collation in case it isn't allowed to have inputs with indeterminate collations, and thus should fail.

In the original wording, it was a bit ambiguous whether the given expression was implicitly not allowed to have inputs with indeterminate collation, or whether this was also evaluated as part of the method (with the latter being the case looking at the underlying implementation).